hibernate-core/src/main/java/org/hibernate/spi/NavigablePath.java

```diff
@@ -75,9 +75,9 @@ public NavigablePath(NavigablePath parent, String localName, @Nullable String al
 		}
 		else {
 			this.localName = localName;
-			this.identifierForTableGroup = StringHelper.isEmpty( parent.getIdentifierForTableGroup() )
+			this.identifierForTableGroup = StringHelper.isEmpty( parent.getFullPath() )
```

Comment:
Calling `getFullPath()` multiple times here is potentially costly as that will recompute a path string on every call. AFAICT from looking at the code, the only difference between `identifierForTableGroup` and `fullPath` is that the latter contains the aliases of the path nodes. By using `getFullPath()` you're essentially adding aliases, which I don't think is correct. The main purpose of `identifierForTableGroup` is to serve as "fetch path" in `FromClauseIndex`, which should IMO be of the form `alias.localName.localName...` i.e. only use the root alias, but following from there, only use local names.

Reply:
> Calling getFullPath() multiple times here is potentially costly as that will recompute a path string on every call

This is only called once on initialization though. If that's still a problem, we can cache the full path value, just like we do for the `identifierForTableGroup` string.

> the only difference between identifierForTableGroup and fullPath is that the latter contains the aliases of the path nodes

That's what's needed for the test case to work, though: the `identifierForTableGroup` of the two `join`s in the query is exactly the same, and without looking at the **parent path's alias** when resolving the `join fetch`es we were wrongly resolving to the same parent table group in `FromClauseIndex`. The "local" path, i.e. the last part of the navigable path, will be without alias as it should to identify the "fetch path".

The joined table groups correspond to the specific aliased joins, retrieving them through the `identifierForTableGroup` when it doesn't contain eventual parent join aliases is IMO incorrect, as all queries with multiple joins with the same path but different aliases will resolve to the same table group.

Reply:
Ok, so AFAIU we need
* `identifierForTableGroup` to be of the form `localName(alias).localName(alias).localName` i.e. skip the last alias
* `fullPath` to be `identifierForTableGroup` + `(alias)`

Right? If that's the case, I would like that `getFullPath()` leverages `identifierForTableGroup` instead of invoking this `fullPathCalculator`.

Reply:
Done, I think that's much cleaner and is the correct form for `identifierForTableGroup`.

I've deprecated the `org.hibernate.spi.NavigablePath.FullPathCalculator` functional interface (and the methods used for it) as it's no longer used. Since this is an SPI change we might not want to backport this to 6.5.